pkg/datapath/linux/devices.go

```diff
@@ -1,521 +1,125 @@
-	return devs
-}
-
-// Exclude devices that have one or more of these flags set.
-var excludedIfFlagsMask uint32 = unix.IFF_SLAVE | unix.IFF_LOOPBACK
-
-// isViableDevice returns true if the given link is usable and Cilium should attach
-// programs to it.
-func (dm *DeviceManager) isViableDevice(l3DevOK, hasDefaultRoute bool, link netlink.Link) bool {
-	name := link.Attrs().Name
-
-	// Do not consider any of the excluded devices.
-	for _, p := range excludedDevicePrefixes {
-		if strings.HasPrefix(name, p) {
-			log.WithField(logfields.Device, name).
-				Debugf("Skipping device as it has excluded prefix '%s'", p)
-			return false
-		}
-	}
-
-	// Skip devices that have an excluded interface flag set.
-	if link.Attrs().RawFlags&excludedIfFlagsMask != 0 {
-		log.WithField(logfields.Device, name).Debugf("Skipping device as it has excluded flag (%x)", link.Attrs().RawFlags)
-		return false
+	hasWildcard := false
+	userDevices := option.Config.GetDevices()
+	for _, name := range userDevices {
+		hasWildcard = hasWildcard || strings.HasSuffix(name, "+")
 	}
 
-	// Ignore L3 devices if we cannot support them.
-	if !l3DevOK && !mac.LinkHasMacAddr(link) {
-		log.WithField(logfields.Device, name).
-			Info("Ignoring L3 device; >= 5.8 kernel is required.")
-		return false
-	}
+	devs, _ := state.ViableDevices(dm.params.DeviceTable.Reader(dm.params.DB.ReadTxn()))
+	names := state.DeviceNames(devs)
 
-	// If user specified devices or wildcards, then skip the device if it doesn't match.
-	if !dm.filter.match(name) {
-		return false
+	if len(names) == 0 && hasWildcard {
+		// Fail if user provided a device wildcard which didn't match anything.
+		return nil, fmt.Errorf("No device found matching %v", userDevices)
 	}
 
-	switch link.Type() {
-	case "veth":
-		// Skip veth devices that don't have a default route.
-		// This is a workaround for kubernetes-in-docker. We want to avoid
-		// veth devices in general as they may be leftovers from another CNI.
-		if !hasDefaultRoute {
-			log.WithField(logfields.Device, name).
-				Debug("Ignoring veth device as it has no default route")
-			return false
-		}
-
-	case "bridge", "openvswitch":
-		// Skip bridge devices as they're very unlikely to be used for K8s
-		// purposes. In the rare cases where a user wants to load datapath
-		// programs onto them they can override device detection with --devices.
-		log.WithField(logfields.Device, name).Debug("Ignoring bridge-like device")
-		return false
+	option.Config.SetDevices(names)
 
-	}
-
-	if link.Attrs().MasterIndex > 0 {
-		if master, err := dm.handle.LinkByIndex(link.Attrs().MasterIndex); err == nil {
-			switch master.Type() {
-			case "bridge", "openvswitch":
-				log.WithField(logfields.Device, name).Debug("Ignoring device attached to bridge")
-				return false
-
-			case "bond", "team":
-				log.WithField(logfields.Device, name).Debug("Ignoring bonded device")
-				return false
+	var nodeDevice *state.Device
+	if k8sEnabled {
+		nodeIP := node.GetK8sNodeIP()
```

Comment:
This requires use of LocalNodeStore and changes in https://github.com/cilium/cilium/pull/24866 to work as NodeIP is otherwise uninitialized.